### PR TITLE
Partial implementation

### PR DESCRIPTION
**Pull request by kashimiz**
Partial implementation:
* Orchestrator parses HistoryEvent array from runtime.
* Aligned output with Durable's out of proc execution schema.
* Added Task-like object for use in orchestrator code.
* Implemented callActivityAsync, createTimer, waitForExternalEvent,
getInput, and currentUtcDateTime.
* Updated samples and README.

### CODE REVIEW

File: src/orchestrator.ts [hunk 1]
```diff
@@ -3,106 +3,361 @@ import * as debug from "debug";
...
+    OrchestratorStarted = 12,
+    OrchestratorCompleted = 13,
+    EventRaised = 14,
+    ContinueAsNew = 15,
+    GenericEvent = 16,
+    HistoryState = 17,
+}
+
+// tslint:disable-next-line:max-classes-per-file
+abstract class Action {
```
Comment (christopheranderson) [inline src/orchestrator.ts]: Abstract class is fine here, but you can also consider using an interface to enforce a public type exiting on an object. If you're not share common functionality, that might be more flexible.

**Comment by christopheranderson** [comment]
Additionally, I would recommend:
 - [ ] Update the tslint version package.json to `^5.9.1` and fix all the issues that come up when you do that. You have a fair amount of "nits" that would be caught by the latest tslint.

File: test/sample/cancel-timer/index.js [hunk 2]
```diff
@@ -0,0 +1,15 @@
+const df = require("../../../lib/");
+const moment = require('moment');
+
+module.exports = df(function*(context){
+    let expiration = moment.utc(context.df.currentUtcDateTime).add(2, 'm');
```
Comment (christopheranderson) [inline test/sample/cancel-timer/index.js]: prefer using const instead of let

**Comment by christopheranderson** [comment]
For the samples. If you check in a local.settings.json which contains:
```json
{
  "IsEncrypted": false,
  "Values": {
    "AzureWebJobsDashboard": "UseDevelopmentStorage=true",
    "AzureWebJobsStorage": "UseDevelopmentStorage=true"
  },
  "ConnectionStrings": {}
}
```

Folks will be able to run the sample without needing to do anything other than install dependencies.

File: README.md [hunk 3]
```diff
@@ -1,296 +1,115 @@
-# Durable Functions for Node.js
+## Durable Functions for Node.js
```
[the remark this answers is not in the document]
Comment (kashimiz) [reply README.md]: ✔

File: README.md [hunk 4]
```diff
@@ -1,296 +1,115 @@
...
 
-Provides a shim to write your Durable Functions orchestration for Node.js
+This library provides a shim to write your [Durable Functions](https://docs.microsoft.com/en-us/azure/azure-functions/durable-functions-overview) orchestrator functions in [Node.js](https://nodejs.org/en/), using Durable's out-of-proc execution protocol. **JS orchestrators have the same [code constraints](https://docs.microsoft.com/en-us/azure/azure-functions/durable-functions-checkpointing-and-replay#orchestrator-code-constraints) as C# orchestrators.**
```
[the remark this answers is not in the document]
Comment (kashimiz) [reply README.md]: Good call.

File: src/orchestrator.ts [hunk 5]
```diff
@@ -3,106 +3,361 @@ import * as debug from "debug";
...
+        public EventType: EventType,
+        public FireAt: Date,                // TimerScheduled, TimerFired
+        public IsPlayed: boolean = false,
+        public Input: any,                  // EventRaised
+        public Name: string,
+        public Result: string,
+        public TaskScheduledId: number,     // TaskCompleted
+        public TimerId: number,             // TimerFired
+        public Timestamp: Date,
+        public IsProcessed: boolean = false // internal use
```
[the remark this answers is not in the document]
Comment (kashimiz) [reply src/orchestrator.ts]: ✔

File: test/sample/cancel-timer/index.js (hunk 2 above)
Comment (kashimiz) [reply test/sample/cancel-timer/index.js]: ✔

File: src/orchestrator.ts [hunk 6]
```diff
@@ -3,106 +3,361 @@ import * as debug from "debug";
...
+        return createdTimer ?
+            state.filter((val: HistoryEvent) => {
+                return val.EventType === EventType.TimerFired
+                    && val.TimerId === createdTimer.EventId;
+            })[0]
+            : undefined;
     }
 }
 
 // tslint:disable-next-line:max-classes-per-file
```
[the remark this answers is not in the document]
Comment (kashimiz) [reply src/orchestrator.ts]: ✔

File: src/orchestrator.ts (hunk 1 above)
Comment (kashimiz) [reply src/orchestrator.ts]: ✔

**Pull request merged**